src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -0,0 +1,443 @@
+// Copyright 2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::impls::gbq::writer::Config;
+use crate::connectors::prelude::*;
+use async_std::prelude::{FutureExt, StreamExt};
+use futures::stream;
+use googapis::google::cloud::bigquery::storage::v1::append_rows_request::ProtoData;
+use googapis::google::cloud::bigquery::storage::v1::big_query_write_client::BigQueryWriteClient;
+use googapis::google::cloud::bigquery::storage::v1::table_field_schema::Type as TableType;
+use googapis::google::cloud::bigquery::storage::v1::{
+    append_rows_request, table_field_schema, write_stream, AppendRowsRequest,
+    CreateWriteStreamRequest, ProtoRows, ProtoSchema, TableFieldSchema, WriteStream,
+};
+use gouth::Token;
+use prost::encoding::WireType;
+use prost_types::{field_descriptor_proto, DescriptorProto, FieldDescriptorProto};
+use std::collections::HashMap;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::metadata::{Ascii, MetadataValue};
+use tonic::service::Interceptor;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::{Request, Status};
+
+pub(crate) struct GbqSink {
+    client: Option<BigQueryWriteClient<InterceptedService<Channel, AuthInterceptor>>>,
+    write_stream: Option<WriteStream>,
+    mapping: Option<JsonToProtobufMapping>,
+    config: Config,
+}
+
+pub(crate) struct AuthInterceptor {
+    token: MetadataValue<Ascii>,
+}
+
+impl Interceptor for AuthInterceptor {
+    fn call(&mut self, mut request: Request<()>) -> ::std::result::Result<Request<()>, Status> {
+        request
+            .metadata_mut()
+            .insert("authorization", self.token.clone());
+
+        Ok(request)
+    }
+}
+
+struct Field {
+    table_type: TableType,
+    tag: u32,
+
+    // ignored if the table_type is not struct
+    subfields: HashMap<String, Field>,
+}
+
+struct JsonToProtobufMapping {
+    fields: HashMap<String, Field>,
+    descriptor: DescriptorProto,
+}
+
+fn map_field(
+    schema_name: &str,
+    raw_fields: &Vec<TableFieldSchema>,
+) -> (DescriptorProto, HashMap<String, Field>) {
+    let mut nested_types = vec![];
```

Comment:
It'd be nice to use Vec::with_capacity() here since we kind of know the size already - this will save allocations.